### PR TITLE
runtime support for repeated task

### PR DESCRIPTION
**Pull request by drunkirishcoder**

### CODE REVIEW

File: framework/src/runtime.rs [hunk 1]
```diff
@@ -23,9 +28,11 @@ impl Runtime {
         info!("initializing context:\n{}", configuration);
         let mut context = initialize_system(configuration)?;
         context.start_schedulers();
+        let tokio_rt = TokioRuntime::new()?;
         Ok(Runtime {
             context,
-            on_signal: Box::new(|_| Err(0)),
+            tokio_rt,
+            on_signal: Arc::new(|_| Err(0)),
```
Comment (zeeshanlakhani) [inline framework/src/runtime.rs]: know it's not exactly the line within this PR, but @drunkirishcoder and I discussed him adding in an example within the comment doc for when a user wants to impl `set_on_signal`.

File: framework/src/runtime.rs [hunk 2]
```diff
@@ -49,21 +56,35 @@ impl Runtime {
     where
         T: Fn(i32) -> std::result::Result<(), i32> + 'static,
     {
-        self.on_signal = Box::new(on_signal);
+        self.on_signal = Arc::new(on_signal);
+    }
+
+    /// Adds a repeated task to run at an interval
+    pub fn add_task_to_run<T>(&mut self, task: T, interval: Duration)
```
Comment (zeeshanlakhani) [inline framework/src/runtime.rs]: @drunkirishcoder do we also want to allow for a task that is not on an interval, like a one off, or run via a timer just once (like run some task 5 seconds from now)? I was just (re?)-reading https://tokio.rs/docs/going-deeper/timers/ and thinking about how some variations may be useful.

File: framework/src/runtime.rs [hunk 3]
```diff
@@ -3,18 +3,23 @@ use common::Result;
 use config::{NetBricksConfiguration, CLI_ARGS};
```
Comment (zeeshanlakhani) [inline framework/src/runtime.rs]: I know we test the runtime out via examples, but maybe we want some simple unit tests around it, including sending signals, using a timer, etc...? Not a biggie, but thinking aloud on it.

Comment (drunkirishcoder) [reply framework/src/runtime.rs]: I'm not sure these things (unix signals, timer behavior, nb context, process exit code) are easily unit testable.

File: framework/src/runtime.rs (hunk 2 above)
Comment (drunkirishcoder) [reply framework/src/runtime.rs]: Can't think of a use case for a run-once behavior right now. The run once tasks we do are usually initialization things that don't need to be scheduled. But if we do encounter an use case, the capability can be pretty easily added to the runtime later.

File: framework/src/runtime.rs (hunk 3 above)
Comment (zeeshanlakhani) [reply framework/src/runtime.rs]: Figured

File: framework/src/runtime.rs (hunk 2 above)
Comment (clinedome) [reply framework/src/runtime.rs]: I think most scheduling systems I've seen have these variations, scheduling a single task after a delay, or scheduling a repeated task after a delay (and with a delay between them). That said, agree with @drunkirishcoder... if we don't need it right now, I think we probably hold off until we do.

File: examples/collect-metrics/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+}
+
+fn main() -> Result<()> {
+    let configuration = load_config()?;
+    println!("{}", configuration);
+    let mut runtime = Runtime::init(&configuration)?;
+    runtime.add_pipeline_to_run(install);
+    runtime.add_task_to_run(|| METRICS.print(), Duration::from_secs(1));
+    runtime.execute()
```
Comment (clinedome) [inline examples/collect-metrics/src/main.rs]: Nice, this initialization process is 👌

File: examples/collect-metrics/src/main.rs [hunk 5]
```diff
@@ -57,7 +54,7 @@ impl Metrics {
 }
 
 lazy_static! {
-    static ref METRICS: Arc<Metrics> = Arc::new(Metrics::new());
+    static ref METRICS: Metrics = Metrics::new();
```
Comment (zeeshanlakhani) [inline examples/collect-metrics/src/main.rs]: 👍

File: framework/src/runtime.rs [hunk 6]
```diff
@@ -52,6 +52,24 @@ impl Runtime {
     /// `Err(i32)` indicates to exit the process with the given exit
     /// code. If no signal handler is provided, the process will terminate
     /// on any signal received.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// let mut runtime = Runtime::init(&CONFIG)?;
```
Comment (zeeshanlakhani) [inline framework/src/runtime.rs]: \o/

File: framework/src/runtime.rs (hunk 2 above)
Comment (drunkirishcoder) [reply framework/src/runtime.rs]: Yep agreed. Also I may have lied about it being easy. I can see more likely you would want the ability to schedule a run-once task from inside a pipeline and there would be some threading issues to work through. nonetheless we will get to it when it comes up.

**Pull request merged**